docs/operation/opa-preloading.md

```diff
@@ -0,0 +1,116 @@
+# OPA Pre-loading Feature
+
+## Overview
+
+This feature implements issue #3526 by moving OPA instance startup and bundle downloading out of filter creation to make route processing faster and more reliable.
+
+## Problem
+
+Previously, OPA filters created instances synchronously during filter creation, which:
+- Blocked route processing during bundle downloads
+- Could cause timeouts with large bundles or slow networks
+- Violated the assumption that filter creation should be fast
+
+## Solution
+
+When enabled via the `--enable-open-policy-agent-preloading` flag, OPA instances are pre-loaded during route processing instead of filter creation:
+
+1. **Route PreProcessor**: Scans routes for OPA filter usage and pre-loads instances
+2. **Non-blocking Filter Creation**: Filters only lookup ready instances, failing fast if not available
+3. **Smart Loading Strategy**: Parallel loading on startup, sequential on route updates
+
+## Usage
+
+### Enable the Feature
+
+```bash
+skipper --enable-open-policy-agent --enable-open-policy-agent-preloading
+```
+
+### Configuration
+
+- `--enable-open-policy-agent-preloading`: Enable the pre-loading behavior (default: false)
+- All existing OPA configuration flags remain the same
+
+### Behavior Changes
+
+#### With Pre-loading Disabled (Default)
+- **Filter Creation**: Synchronous, blocks until OPA instance is ready
+- **Route Processing**: Fast, instances already created
+- **Backward Compatibility**: 100% compatible with existing behavior
+
+#### With Pre-loading Enabled
```

Comment:
it might help to explain how new routes would be added, by being picked up be the route processing (that runs regularly) which triggers the instance creation and the filter might only become active after the second route processing, as instance creation is async (if I understand the process correctly)

Reply:
I added a log line at https://github.com/zalando/skipper/blob/master/routing/datasource.go#L612 to confirm if pre processors are getting called for regular route update check and it was only printed if there was an effective change to route definitions in .eskip file. In that sense a failed OPA route will not become active again, unless there is a route modification. I understand it could have been an overhead if it indefinitely keep on re-trying with pre processor getting called periodically as well.

@AlexanderYastrebov @szuecs could you shed some light on expected execution of pre-processors for route updates and whether invalid routes get re-attempted.

Reply:
That's correct, but I think autoscaling will make sure it will get triggered more often in production clusters that you may think.

Reply:
We had an offline discussion and concluded to have a retry go routine, attempting failed instances either by deducing it using required instances vs available instances or tracking the instance status in OPA registry.